client-sdk/ts-web/rt/src/mrae.ts

```diff
@@ -0,0 +1,53 @@
+import * as oasis from '@oasisprotocol/client';
+import * as ed from '@noble/ed25519';
+import {sha512_256} from 'js-sha512';
+
+var deoxysii = require('deoxysii');
+
+const BoxKDFTweak = 'MRAE_Box_Deoxys-II-256-128';
+
+/**
+ * deriveSymmetricKey derives a MRAE AEAD symmetric key suitable for use with the asymmetric
+ * box primitives from the provided X25519 public and private keys.
+ */
+
+export function deriveSymmetricKey(publicKey: Uint8Array, privateKey: Uint8Array): Uint8Array {
+    const pmk = ed.curve25519.scalarMult(oasis.misc.toHex(privateKey), oasis.misc.toHex(publicKey));
+    var kdf = sha512_256.hmac.create(BoxKDFTweak);
+    kdf.update(pmk);
+    return oasis.misc.fromHex(kdf.hex());
+}
+
+/**
+ * boxSeal boxes ("seals") the provided additional data and plaintext via
+ * Deoxys-II-256-128 using a symmetric key derived from the provided
+ * X25519 public and private keys.
+ */
+export function boxSeal(
+    nonce: Uint8Array,
+    plainText: Uint8Array,
+    associateData: Uint8Array,
+    publicKey: Uint8Array,
+    privateKey: Uint8Array,
+): Uint8Array {
+    const sharedKey = deriveSymmetricKey(publicKey, privateKey);
+    var aead = new deoxysii.AEAD(sharedKey);
```

Comment:
This also has a non-trivial setup cost such that the cipher could (should) be cached. It's worse because the JS DeoxysII is not hw accelerated. For consistency with the other impls, though, I won't make a fuss about this. This is also not the lowest hanging fruit–just annoying.

Reply:
how much setup cost is there in this setup? I see from the syntax that this creates an object, which must have some kind of state. is the setup separable from the lifetime of this state at all?

if it is, maybe we could expose something like nacl's 'box before' and 'box/open after' system https://www.npmjs.com/package/tweetnacl/v/1.0.3#naclboxbeforetheirpublickey-mysecretkey